### ISSUE TITLE
Can't use `forwardRef` in TypeScript code that is compiled to dart

### ISSUE THREAD

**Issue by jelbourn** [angular/angular#2127]
## Solution

Have ts2dart remove the `forwardRef` statements so that `forwardRef(()=>MyType)` becomes `MyType`.

example:

```javascript
directives: [
  bind(forwardRef(() => MyType).toClass(forwardRef(() => MyImpl),
  forwardRef(() => MyImpl)
]
```

becomes

```javascript
directives: [
  bind(MyType).toClass(MyImpl),
  MyImpl
]
```

## Problem

I have a component called `MdDialogContainer`. It contains in its template an `MdDialogContent`:
```js
@Component({
  selector: 'md-dialog-container'
})
@View({
  directives: [MdDialogContent]
})
class MdDialogContainer { ... }
```

`MdDialogContent`, however, needs to inject its parent `MdDialogContainer` so that it can set a property on the container:
```js
@Directive({selector: 'md-dialog-content'})
class MdDialogContent {
  constructor(@Parent() dialogContainer: MdDialogContainer, elementRef: ElementRef) {
    dialogContainer.contentRef = elementRef;
  }
}
```
This won't work with the way that hoisting of `class` constructs works in ES6. We can work around this by using a forward declaration for the container:
```js
@Component({
  selector: 'md-dialog-container',
  hostListeners: {
    'body:^keydown': 'documentKeypress($event)'
  }
})
@View({
  templateUrl: 'angular2_material/src/components/dialog/dialog.html',
  directives: [forwardRef(() => MdDialogContent)]
})
class MdDialogContainer { ... }
```

This works great for TS / JS, but is no good when compiled to dart since `directives` isn't `const` any more. 

`@Query` might have been a workaround to this issue, but the `onChange` of the `MdDialogContainer` is too late to have the needed property set (results of `@Query` are not ready yet in the constructor).

This is blocking moving this dialog component to TypeScript.

FYI @mprobst @alexeagle @rkirov

**Comment by mprobst** [comment]
Can you make `forwardRef` a `@const` function?

**Comment by jelbourn** [comment]
All of the arguments to a `const` function have to also be const. I think that the purpose of `forwardRef` and the concept of `const` functions are not compatible.

**Comment by mprobst** [comment]
Wouldn't the class definition be const in Dart, i.e. fully resolved before
runtime?

Jeremy Elbourn <notifications@github.com> schrieb am Di., 26. Mai 2015 um
10:16 Uhr:

**Comment by jelbourn** [comment]
Doesn't seem that way when I tried it, but I don't know much about dart.

@rkirov do you have any insights?

**Comment by rkirov** [comment]
Yeah, this will be a hard one. Dart considers any closure as a non-const value, so `const [() => 3]` is an error. Also any function invocation (except const constructor) is forbidden, so `id(x) => x; const[id(3)]` is also an error.

**Comment by alexeagle** [comment]
@vsavkin says we could just strip the forwardRef function call and unwrap the closure for ts2dart.

**Comment by mhevery** [comment]
Dart does not have a forwardRef problem, and so you can just strip it out.

`forwardRef(()=>MyType)` becomes `MyType`.

example:

```javascript
directives: [
  bind(forwardRef(() => MyType).toClass(forwardRef(() => MyImpl),
  forwardRef(() => MyImpl)
]
```

becomes

```javascript
directives: [
  bind(MyType).toClass(MyImpl),
  MyImpl
]
```

**Comment by mprobst** [comment]
File an issue, please?

Miško Hevery <notifications@github.com> schrieb am Fr., 29. Mai 2015 um
09:21 Uhr:

**Comment by mhevery** [comment]
https://github.com/angular/ts2dart/issues/177

**Comment by jelbourn** [comment]
Fixed via d5236133297eab13ea70b5577d8aa4f7ae1be62c
